### PR TITLE
Copilot-based TS refactors

### PR DESCRIPTION
**Pull request by sandersn**
This PR adds copilot suggestions to various Typescript refactors and quickfixes. It also adds a separate quickfix for implicit any that lets Copilot infer types from the body of a function, which is less context than the TS codefix. However, it creates new interfaces with reasonable names and uses them where appropriate, which the TS codefix cannot.

The existing refactor/quickfixes are followed by Copilot's inline chat now:

- Extract to X now asks copilot to provide a name for the new constant/function/type instead of showing the rename UI.
- Implement X now asks copilot to provide reasonable implementations for inherited interface or abstract methods. This replaces the TS-generated `throw new Error("Not implemented")`.

I think the the UX and performance are both good enough for insider use at this point, so I think it's worth merging behind the `typescript.experimental.aiQuickFix` flag.

Some future work:

- Could the codefix could be skipped if there are lots of types around already?
- Could the codefix prompt by modified to omit "Please generate new interfaces" if there lots of interfaces already declared?
- Fix code formatting!

Extends #190690

Question and work from that PR that are now resolved:

- Infer from types now works in JS files.
- When given enough context, inline chat works well, especially since it will edit code if the AI produces a code sample. I tried showing multiple name suggestions in the chat panel, but I felt like it didn't add enough utility -- and the suggestions were usually all good or all bad, so taking the first suggestion works well enough.

### CODE REVIEW

**Comment by sandersn** [comment]
@mjbvz and @jrieken can you take a look at this?

**Comment by jrieken** [comment]
@sandersn There are some eslint failures. You can see them when running `yarn eslint` (actually the precommit hooks should have reported them...)

I wonder if we should grow the `typescript.experimental.aiQuickFix` setting, e.g this isn't just quick fixes anymore and also more nuanced. Maybe a new setting like `typescript.exprimental.aiCodeActions` which value is a complex object would be best, e.g each key would correspond to a feature

**Comment by sandersn** [comment]
Ah, that's right. I completely messed up the setup of vscode's formatter, so I skipped the precommit hook. I'll go back and figure that out so the machine can fix the formatting for me.

**Comment by sandersn** [comment]
OK, I added a more complex options object. If you just create the top-level object as `{}`, all the code actions are enabled. Then you can set individual ones to false.

I don't know if this is the usual way to do it, so please let me know if there's a standard I should follow.

File: extensions/typescript-language-features/package.json [hunk 1]
```diff
@@ -145,11 +145,58 @@
...
-          "scope": "resource"
+        "typescript.experimental.aiCodeActions": {
+          "type": "object",
+          "default": {},
+          "description": "%typescript.experimental.aiCodeActions%",
+          "scope": "resource",
+          "properties": {
+            "classIncorrectlyImplementsInterface": {
+              "type": "boolean",
+              "default": true,
```
Comment (jrieken) [inline extensions/typescript-language-features/package.json:156]: We aren't so brave and have them off by default

File: extensions/typescript-language-features/package.json [hunk 2]
```diff
@@ -145,11 +145,58 @@
...
+            },
+            "extractType": {
+              "type": "boolean",
+              "default": true,
+              "description": "%typescript.experimental.aiCodeActions.extractType%"
+            },
+            "extractInterface": {
+              "type": "boolean",
+              "default": true,
+              "description": "%typescript.experimental.aiCodeActions.extractInterface%"
```
[the remark this answers is not in the document]
Comment (sandersn) [reply extensions/typescript-language-features/package.json:197]: Oops, I hadn't internalised the setup for configs, and missed the update to package.nls.json

**Review by jrieken — DISMISSED** [review]

**Comment by sandersn** [comment]
I realised that the config still effectively defaulted to true, and that the altered descriptions weren't working.
I fixed that and also created custom descriptions, so please take a look at those for wording.

**Review by jrieken — APPROVED** [review]

**Review by alexr00 — APPROVED** [review]

**Comment by jrieken** [comment]
Merging caused a build failure down the line (when doing the product build...) A revert PR is up (#193810) and this needs to be reopened and looked at again

**Pull request merged**